packages/jupyter-ai/jupyter_ai/chat_handlers/tools.py

```diff
@@ -0,0 +1,260 @@
+                content = file.read()
+                tree = ast.parse(content)
+                tools = []
+                for node in ast.walk(tree):
+                    if isinstance(node, ast.FunctionDef):
+                        for decorator in node.decorator_list:
+                            if (
+                                isinstance(decorator, ast.Name)
+                                and decorator.id == "tool"
+                            ):
+                                tools.append(node.name)
+            return tools  # this is a list
+        except FileNotFoundError as e:  # to do
+            self.reply(f"Tools file not found at {tools_file_path}.")
+
+    def toolChat(self, query):
+        print("TOOL CHAT", query)
+        for chunk in self.app.stream(
+            {"messages": [("human", query)]}, stream_mode="values"
+        ):
+            response = chunk["messages"][-1].pretty_print()
+        return response
+
+    ##### MAIN FUNCTION #####
+    def useLLMwithTools(self, query):
+        """
+        LangGraph documentation : https://langchain-ai.github.io/langgraph/tutorials/introduction/
+        The code below:
+        1. Extracts the function names in the custom tools file
+        2. Adds the tools to the Tool Node
+        3. Binds the Tool Node to the LLM
+        4. Sets up a basic LangGraph with nodes and edges
+        5. Compiles the graph into a runnable app
+        6. This function is then called with a prompt
+        Every time a query is submitted the langgraph is rebuilt in case the tools file has been changed.
+        """
+
+        # Calls the requisite tool in the LangGraph
+        def call_tool(state: MessagesState):
+            messages = state["messages"]
+            response = self.model_with_tools.invoke(messages)
+            return {"messages": [response]}
+
+        # Get all tool objects from the tool files
+        def getTools(file_paths):
+            if len(file_paths) > 0:
+                tool_names = []
+                for file_path in file_paths:
+                    with open(file_path) as file:
+                        exec(file.read())
+                    tool_names = tool_names + self.getToolNames(file_path)
+                tools = [eval(j) for j in tool_names]
+                return tools
+
+        # Get tool file(s), then tools within tool files, and create tool node from tools
+        file_paths = self.getToolFiles(self.tools_file_path)
+        tools = getTools(file_paths)
+        tool_node = ToolNode(tools)
+
+        # Bind tools to LLM
+        # Check if the LLM class takes tools else advise user accordingly.
+        # Can be extended to include temperature parameter
+        try:
+            self.model_with_tools = self.llm.__class__(
+                model_id=self.llm.model_id
+            ).bind_tools(tools)
+        except Exception as e:
+            self.reply(f"Not a chat model, cannot be used with tools. {e}")
```

Comment:
The `try/except` block can be replaced with a preemptive check that raises an exception if the current LLM is not a chat model:

```py
if not self.llm.is_chat_provider:
    raise ExceptionNotChatModel()

self.model_with_tools = self.llm.__class__(
                model_id=self.llm.model_id
            ).bind_tools(tools)
```

Reply:
This custom exception handling also added.